cpp/cmake_modules/ThirdpartyToolchain.cmake

```diff
@@ -582,39 +582,43 @@ else()
   if(ARROW_BOOST_USE_SHARED)
     # Find shared Boost libraries.
     set(Boost_USE_STATIC_LIBS OFF)
+    # Stash BUILD_SHARED_LIBS
     set(BUILD_SHARED_LIBS_KEEP ${BUILD_SHARED_LIBS})
     set(BUILD_SHARED_LIBS ON)
 
     if(MSVC)
       # force all boost libraries to dynamic link
       add_definitions(-DBOOST_ALL_DYN_LINK)
     endif()
-
-    if(ARROW_BOOST_HEADER_ONLY)
-      find_package(Boost REQUIRED)
-    else()
-      find_package(Boost COMPONENTS regex system filesystem REQUIRED)
-      set(BOOST_SYSTEM_LIBRARY Boost::system)
-      set(BOOST_FILESYSTEM_LIBRARY Boost::filesystem)
-      set(BOOST_REGEX_LIBRARY Boost::regex)
-      set(ARROW_BOOST_LIBS ${BOOST_SYSTEM_LIBRARY} ${BOOST_FILESYSTEM_LIBRARY})
-    endif()
-    set(BUILD_SHARED_LIBS ${BUILD_SHARED_LIBS_KEEP})
-    unset(BUILD_SHARED_LIBS_KEEP)
   else()
     # Find static boost headers and libs
     # TODO Differentiate here between release and debug builds
     set(Boost_USE_STATIC_LIBS ON)
-    if(ARROW_BOOST_HEADER_ONLY)
-      find_package(Boost REQUIRED)
+  endif()
+
+  if(ARROW_BOOST_HEADER_ONLY)
+    find_package(Boost REQUIRED)
+  else()
+    find_package(Boost COMPONENTS regex system filesystem REQUIRED)
+    if(CMAKE_VERSION VERSION_LESS 3.5)
+      # Cmake's FindBoost started exporting IMPORTED in 3.5
+      set(BOOST_SYSTEM_LIBRARY ${Boost_SYSTEM_LIBRARY})
+      set(BOOST_FILESYSTEM_LIBRARY ${Boost_FILESYSTEM_LIBRARY})
+      set(BOOST_REGEX_LIBRARY ${Boost_REGEX_LIBRARY})
     else()
-      find_package(Boost COMPONENTS regex system filesystem REQUIRED)
       set(BOOST_SYSTEM_LIBRARY Boost::system)
       set(BOOST_FILESYSTEM_LIBRARY Boost::filesystem)
       set(BOOST_REGEX_LIBRARY Boost::regex)
-      set(ARROW_BOOST_LIBS ${BOOST_SYSTEM_LIBRARY} ${BOOST_FILESYSTEM_LIBRARY})
     endif()
+    set(ARROW_BOOST_LIBS ${BOOST_SYSTEM_LIBRARY} ${BOOST_FILESYSTEM_LIBRARY})
   endif()
+
+  if(ARROW_BOOST_USE_SHARED)
```

Comment:
I'm not sure why this was like this, if someones wants to clarify, I'll gladly add a comment to the PR (@xhochy, @wesm).

Reply:
This is needed to work with `BoostConfig.cmake` provided by Boost (not CMake).
The Boost package provided by MSYS2 installed `BoostConfig.cmake` at the time.
But the current Boost package doesn't install `BoostConfig.cmake`: https://github.com/msys2/MINGW-packages/pull/5326/files

See also: https://github.com/msys2/MINGW-packages/issues/5233

There is no platform that needs this. So we can remove this for now but we may need this again in the future.